server/src/main/java/io/spine/server/delivery/Delivery.java

```diff
@@ -0,0 +1,496 @@
+                            .setStrategy(strategy)
+                            .build();
+        delivery.subscribe(new LocalDispatchingObserver());
+        return delivery;
+    }
+
+    /**
+     * Delivers the messages put into the shard with the passed index to their targets.
+     *
+     * <p>At a given moment of time, exactly one application node may serve messages from
+     * a particular shard. Therefore, in scope of this delivery, an approach based on pessimistic
+     * locking per-{@code ShardIndex} is applied.
+     *
+     * <p>In case the given shard is already processed by some node, this method does nothing.
+     *
+     * <p>The content of the shard is read and delivered on page-by-page basis. The runtime
+     * exceptions occurring while a page is being delivered are accumulated and then the first
+     * exception is rethrown, if any.
+     *
+     * <p>After all the pages are read, the delivery process is launched again for the same shard.
+     * It is required in order to handle the messages, that may have been put to the same shard
+     * as an outcome of the first-wave messages.
+     *
+     * <p>Once the shard has no more messages to deliver, the delivery process ends, releasing
+     * the lock for the respective {@code ShardIndex}.
+     *
+     * @param index
+     *         the shard index to deliver the messages from.
+     */
+    @SuppressWarnings("WeakerAccess")   // a part of the public API.
+    public void deliverMessagesFrom(ShardIndex index) {
+        NodeId currentNode = ServerEnvironment.instance()
+                                              .nodeId();
+        Optional<ShardProcessingSession> picked =
+                workRegistry.pickUp(index, currentNode);
+        if (!picked.isPresent()) {
+            return;
+        }
+        ShardProcessingSession session = picked.get();
+        try {
+            int deliveredMsgCount;
+            do {
+                deliveredMsgCount = doDeliver(session);
+            } while (deliveredMsgCount > 0);
+        } finally {
+            session.complete();
+        }
+    }
+
+    /**
+     * Runs the delivery for the shard, which session is passed.
+     *
+     * Returns the number of delivered messages.
+     */
+    private int doDeliver(ShardProcessingSession session) {
+        ShardIndex index = session.shardIndex();
+        Timestamp now = Time.currentTime();
+        Timestamp idempotenceWndStart = Timestamps.subtract(now, idempotenceWindow);
+
+        Page<InboxMessage> startingPage = inboxStorage.readAll(index);
+        Optional<Page<InboxMessage>> maybePage = Optional.of(startingPage);
+
+        int totalMessagesDelivered = 0;
+        while (maybePage.isPresent()) {
+            Page<InboxMessage> currentPage = maybePage.get();
+            ImmutableList<InboxMessage> messages = currentPage.contents();
+            if (messages.isEmpty()) {
+                maybePage = currentPage.next();
+                continue;
```

Comment:
We could get rid of this `continue` instruction. Just invert the condition and extract
```java
maybePage = currentPage.next();
```
out of the `if` statement, as both branches end with this statement.